src/pages/RealmExplorer.js

```diff
@@ -3,201 +3,214 @@
+        .map(([field, type]) => ({
+          [field]: parseType(type),
+        }))
+        .reduce((acc, field) => ({ ...acc, ...field }), {});
+      return { [name]: fields };
+    })
+    .reduce((acc, object) => ({ ...acc, ...object }));
+
+const OBJECT_TYPES = getObjectTypes(schema);
+const OBJECT_FIELDS = getObjectFields(schema);
+
+const toCapitalCase = value => value.charAt(0).toUpperCase() + value.slice(1);
+
+const parseString = value => String(value);
+const parseNumber = value => String(value);
+const parseObject = value => value.id;
+const parseArray = value => value.length;
+const parseBoolean = value => toCapitalCase(String(value));
+const parseDate = value => value.toString();
+
+const parseCell = (value, type) => {
+  if (value === null || value === undefined) return 'N/A';
+  switch (type) {
+    case 'string':
+      return parseString(value);
+    case 'number':
+      return parseNumber(value);
+    case 'object':
+      return parseObject(value);
+    case 'array':
+      return parseArray(value);
+    case 'boolean':
+      return parseBoolean(value);
+    case 'date':
+      return parseDate(value);
+    default:
+      return '';
   }
+};
 
-  onFilterChange(filterString) {
-    const { dataSource } = this.state;
-
-    if (!this.unfilteredData) return;
-
-    let data = null;
-    if (filterString === '') {
-      // if filter is emptied, clear filter.
-      data = this.unfilteredData;
-    } else {
-      try {
-        // Use |this.unfilteredData| to avoid stacking filters.
-        data = this.unfilteredData.filtered(filterString);
-      } catch (err) {
-        // Ignore error silently.
-      }
-    }
+const getInitialState = database => {
+  const objectString = 'Requisition';
+  const objectData = database.objects(objectString);
+  const searchString = '';
+  const filterString = '';
+  const filteredData = objectData;
+  return { objectString, objectData, searchString, filterString, filteredData };
+};
 
-    if (data) {
-      this.setState({
-        data,
-        dataSource: dataSource.cloneWithRows(data),
-      });
+const getUpdatedState = (database, state) => {
+  const { objectString, objectData, searchString, filterString, filteredData } = state;
+  const newState = { searchString, filterString };
+  const updateObject = searchString !== objectString && OBJECT_TYPES.indexOf(searchString) >= 0;
+  newState.objectString = updateObject ? searchString : objectString;
+  newState.objectData = updateObject ? database.objects(newState.objectString) : objectData;
+  if (filterString === '') {
+    newState.filteredData = newState.objectData;
+  } else {
+    try {
+      newState.filteredData = newState.objectData.filtered(filterString);
+    } catch (err) {
+      newState.filteredData = updateObject ? newState.objectData : filteredData;
     }
   }
 
-  onSearchChange(searchTerm) {
-    const { database } = this.props;
-    const { dataSource } = this.state;
+  return newState;
```

Comment:
SRP - personally I think this is doing to much

```
const filterData = (newFilterString, state) => {
    const { objectData } = state;
    try...
    const newData = objectData.filtered(filterString)

    catch..
    const newData = objectData

    return {...state, objectData: newData }
}
```

and a similar method for `searchString`


Then, in the component

```
const onFilterData = useCallback((value) => setState((state) => filterData(value)), []);
```

can pass a function to `setState` and it will inject the current `state`

Reply:
That's much nicer. Just to clarify, did you mean to put `[state.filterString]` as the second arg of `useCallback`? Also think that `filterData(value)` should be `filterData(value, state)`?

Reply:
Note: `setState` here would be index 1 return from your `useState` call cant remember what it was hehe

if you pass a function to the `setState` return value, the current state gets injected for you https://reactjs.org/docs/hooks-reference.html#usestate

so:

- Function being passed to your `setState`: `oldState => filterData(value)`
- Function being passed into `useCallback`: `value => setState(oldState => filterData(value))`

The dependent here, is only `setState` - but I *think* you can use the initial value of this from the closure, not 100% but if  so, it can be fully memoized. 

You could technically do some funny things with functions to keep a constant reference, with a closure to `setState`, but more understandable to just memoize I think.

Reply:
Agree 100%. a7496e193ca5ad1349e01923bbc2387fba3c02d3.